lib/ratelimit.c

```diff
@@ -63,3 +63,50 @@ tb_ratelimit_allow(struct token_bucket_ratelimit_state *tbrs)
 
 	return rc;
 }
+
+void
+log_ratelimit_state_init(struct log_ratelimit_state *lrs,
+	uint32_t ratelimit_level, uint32_t interval, uint32_t burst)
+{
+	lrs->ratelimit_level = ratelimit_level;
+	lrs->interval = interval;
+	lrs->burst = burst;
+	lrs->printed = 0;
+	lrs->missed = 0;
+	lrs->begin = 0;
+}
+
+/*
+ * Rate limiting log entries.
+ *
+ * Returns:
+ * - true means go ahead and do it.
+ * - false means callbacks will be suppressed.
+ */
+bool
+log_ratelimit_allow(struct log_ratelimit_state *lrs)
+{
+	uint64_t now;
+
+	if (!lrs->interval)
+		return true;
+
+	now = rte_rdtsc();
+
+	if (!lrs->begin)
+		lrs->begin = now;
+
+	if (lrs->begin + lrs->interval * cycles_per_ms < now) {
+		lrs->begin = now;
+		lrs->printed = 0;
+		lrs->missed = 0;
+	}
+
+	if (lrs->burst && lrs->burst > lrs->printed) {
```

Comment:
We can drop the first part of the test expression: `lrs->burst &&`. The reason for this is that, if `burst` is not zero, the first part of the expression doesn't make a difference. If `burst` is zero, `lrs->burst > lrs->printed` too is always false because field `printed` is of type `uint32_t` and zero cannot be bigger than any other number `printed` can have.